Use OpenAPI V3 for client side SMP

Jefftree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Use OpenAPI V3 for client side apply

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Client-side apply will use OpenAPI V3 by default
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Jefftree committed:
Use OpenAPI V3 for client side SMP

Jefftree committed:
use OpenAPIV3 for kubectl diff

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -182,6 +196,92 @@ func (p *Patcher) buildMergePatch(original, modified, current []byte) ([]byte, e
+
+// TODO: GVK parser should be moved elsewhere
+// Is this duplicated somewhere?
+func gvkMatches(targetGVK schema.GroupVersionKind, ext spec.Extensions) bool {
```
There's a strong possibility that Alex has already written a piece of code to do that, I also suspect maybe sean has written a layer that goes on top of the OpenAPI client to find the object directly.

Jefftree replied:
I saw Sean's snippet in query params verifier v3. It's really concise and I'm going to opt to duplicate it because patch requires checking for two different kinds of gvks. Schemas can have a list of gvks while the patch endpoint can only have a single gvk.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -118,7 +122,17 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, namespace, na
-	if p.OpenapiSchema != nil {
+	if p.OpenAPIV3Root != nil {
+		// TODO: Capture error and log maybe?
+		if gvkSupported, _ := p.gvkSupportsPatchOpenAPIV3(p.Mapping.GroupVersionKind); gvkSupported {
+			patch, err = p.buildStrategicMergePatchFromOpenAPIV3(original, modified, current)
```
it seems like these two things are running together yet they have a lot of common code? Can we actually return the `PatchMeta` if we find one?

Jefftree replied:
I like keeping the two separate because they look at different components of the OpenAPI. `gvkSupportsPatchOpenAPIV3` looks at the PATCH verb for a resource path (does not look at schema at all) while `buildStrategicMergePatchFromOpenAPIV3` looks at Component definitions and traverses the schema resource.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -118,7 +122,17 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, namespace, na
+		}
+	}
+
+	if patch == nil && p.OpenapiSchema != nil {
```
One problem I have with this code is that `p.OpenapiSchema` has already downloaded the OpenAPI v2. I think there's a different interface somewhere that allows a lazy fetching of OpenAPI v2. Otherwise we will always download both in apply.

Jefftree replied:
Not yet, we'll be reviving https://github.com/kubernetes/kubernetes/pull/116565 to lazy fetch OpenAPI V2, but I think we can proceed on this in parallel? And yes, this will currently download both v2 and v3 which the linked PR should fix.

Jefftree commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -148,6 +150,89 @@ func GetTagStructTypeOrDie(dataStruct interface{}) reflect.Type {
+	if refString := l.Schema.Ref.String(); refString != "" {
+		str := strings.TrimPrefix(refString, "#/components/schemas/")
+		sch, ok := l.SchemaList[str]
+		if ok {
```
`fmt.Errorf("should not happen")` :)

seans3 commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -182,6 +196,92 @@ func (p *Patcher) buildMergePatch(original, modified, current []byte) ([]byte, e
+	if ext == nil {
+		return false
+	}
+	f, ok := ext["x-kubernetes-group-version-kind"]
```
Is there a constant somewhere for this string (`x-kubernetes-group-version-kind`)?

Jefftree replied:
There is in `util/openapi` in kubectl but that's coupled with OpenAPI V2 and I was hoping we can keep OpenAPI V2 and V3 coupling to a minimum. I declared a constant in this file instead.

seans3 reviewed (commented):
I think this approach looks good. I would hope for more `kubectl` tests.

Jefftree commented:
@seans3 I added kubectl testing by extending the existing apply tests and showing that OpenAPI V3 produces the same result. Let me know if there are other cases you'd like to see.

Jefftree commented:
/triage accepted

apelisse commented on staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/meta.go:
```diff
@@ -148,6 +150,89 @@ func GetTagStructTypeOrDie(dataStruct interface{}) reflect.Type {
+		return PatchMetaFromOpenAPIV3{}, nil
+	}
+	if len(s.Schema.Properties) == 0 {
+		return PatchMetaFromOpenAPIV3{}, fmt.Errorf("unable to find api field \"%s\"", key)
```
Should you look at additionalProperties?

Jefftree replied:
Does it make a difference to look at additionalProperties? It's only necessary if we have a complex map and need to traverse further to look for merge/patch options. k/k generally uses association lists for more complex types which do not touch additionalProperties. k/k has simple maps (eg: `map[string]string`) which do not need to recurse to the type of the map.

apelisse replied:
If we don't use additionalProperties in k8s then yeah, it's fine, but I don't know if that's 100% true?

Jefftree replied:
We use additionalProperties for simple maps in k/k, mainly `map[string]string` which is covered by test cases involving SimpleMap (https://github.com/kubernetes/kubernetes/blob/31140a5f97f89495ddfd748dc4d283832da672eb/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go#L108)

apelisse replied:
Yeah, I couldn't find any in k8s.io/api, we're probably good, thanks.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -283,6 +286,12 @@ func (flags *ApplyFlags) ToOptions(f cmdutil.Factory, cmd *cobra.Command, baseNa
+	openAPIV3Client, err := f.OpenAPIV3Client()
+	if err == nil {
+		cachedOpenAPIV3Client := cachedopenapi.NewClient(openAPIV3Client)
+		openAPIV3Root = openapi3.NewRoot(cachedOpenAPIV3Client)
```
Do we need to get the root here?

Jefftree replied:
NewRoot creates the interface for fetching the OpenAPI V3 documents. It doesn't make a call to the apiserver. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/openapi3/root.go#L69

apelisse replied:
Thanks, that's great

apelisse commented:
How would we disable the feature if this doesn't work though? Like if openapi v3 generates the wrong patch, is there a way to revert back to v2?

Jefftree committed:
Add feature toggle for OpenAPI V3 apply in kubectl

Jefftree commented:
Added a feature flag for OpenAPI V3 SMP Apply that is on by default. I've added a test that tests disabling the feature gate will result in the OpenAPI V3 getter being completely skipped and also made sure every apply test passes regardless of whether OpenAPI V3 Apply feature flag is toggled on or off. 

Tests indent levels were modified for the toggle so I'd recommend reviewing with whitespace indentations off 

![image](https://github.com/kubernetes/kubernetes/assets/7691399/a2c10f8a-f23c-4a6e-aac6-f4dc7e9b633c)

Jefftree committed:
Lazy load OpenAPIV2

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -283,6 +286,12 @@ func (flags *ApplyFlags) ToOptions(f cmdutil.Factory, cmd *cobra.Command, baseNa
+	var openAPIV3Root openapi3.Root
+	openAPIV3Client, err := f.OpenAPIV3Client()
+	if err == nil {
+		cachedOpenAPIV3Client := cachedopenapi.NewClient(openAPIV3Client)
```
Probably not for now, but we should probably have the factory return a cachedopenapi.Client, that's already what we do for v2. WDYT @seans3

Jefftree replied:
It's a fairly simple change, I changed the factory to use a cached openapi v3 client.

seans3 replied:
Sounds good (on the `cachedopenapi.Client`).

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -222,6 +222,7 @@ func (p *Patcher) buildMergePatch(original, modified, current []byte) ([]byte, e
 // gvkSupportsPatchOpenAPIV3 checks if a particular GVK supports the patch operation.
 // It returns an error if the OpenAPI V3 could not be downloaded.
 func (p *Patcher) gvkSupportsPatchOpenAPIV3(gvk schema.GroupVersionKind) (bool, error) {
+	// Bypassing root to save apiserver memory.
```
That doesn't by-pass the root though. And I think that's fine. We usually apply a lot more than 1 resource types, going through the root might avoid a LOT with the cache busting.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -118,14 +130,44 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, namespace, na
+			// with OpenAPI V3 not present in V2.
+			klog.V(5).Infof("warning: OpenAPI V3 path does not exist - group: %s, version %s, kind %s\n",
+				p.Mapping.GroupVersionKind.Group, p.Mapping.GroupVersionKind.Version, p.Mapping.GroupVersionKind.Kind)
+		} else {
+			if gvkSupported {
```
nit: else if
```suggestion
		} else if gvkSupported {
```

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -118,14 +130,44 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, namespace, na
+					patchType = types.StrategicMergePatchType
+				}
+			} else {
+				klog.V(5).Infof("warning: OpenAPI V3 path does not support strategic merge patch - group: %s, version %s, kind %s\n",
```
Should we really have a warning about this? This just means it's a CRD?

Jefftree replied:
We shouldn't warn, hence the v5 info log. I'd like to make sure every case is logged in case these rare cases do come up and users can provide us with the full log for better debugging. lmk if you think this should be a different log level or removed completely.

apelisse replied:
Yeah, if that's super mild then that's fine thanks.

apelisse commented:
Only minor nits, feel free to push back.

Jefftree commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -283,6 +286,12 @@ func (flags *ApplyFlags) ToOptions(f cmdutil.Factory, cmd *cobra.Command, baseNa
 	openAPISchema, _ := f.OpenAPISchema()
+	var openAPIV3Root openapi3.Root
+	openAPIV3Client, err := f.OpenAPIV3Client()
+	if err == nil && !cmdutil.OpenAPIV3Apply.IsDisabled() {
```
Added a warning. Not really sure about the log level, I see most commands use v4 and stuck with that

seans3 commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply_test.go:
```diff
@@ -63,12 +65,19 @@ import (
+	fakeSchema          = sptest.Fake{Path: filepath.Join("..", "..", "..", "testdata", "openapi", "swagger.json")}
+	fakeOpenAPIV3Legacy = sptest.OpenAPIV3Getter{Path: filepath.Join("..", "..", "..", "testdata", "openapi", "v3", "api", "v1.json")}
+	fakeOpenAPIV3AppsV1 = sptest.OpenAPIV3Getter{Path: filepath.Join("..", "..", "..", "testdata", "openapi", "v3", "apis", "apps", "v1.json")}
+	// testingOpenAPISchemas = []testOpenAPISchema{FakeOpenAPISchema}
```
Should we remove this commented out line?

Jefftree replied:
Fixed.

Jefftree commented:
I'll send a follow up PR to for linter hints since it involves changing invocations of existing tests. It's a non blocking CI.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/util/factory_client_access.go:
```diff
@@ -214,5 +214,5 @@ func (f *factoryImpl) OpenAPIV3Client() (openapiclient.Client, error) {
 	}
 
-	return discovery.OpenAPIV3(), nil
+	return cached.NewClient(discovery.OpenAPIV3()), nil
```
clean 👍

apelisse commented:
/lgtm
/approve
